### PR TITLE
[In progress] Logger argument support from run settings

### PR DESCRIPTION
**Pull request by abhishekkumawat23**
**Change 1**:
From command line, loggers are given as:
/logger:trx /logger:blame;key1=value1

Enabling support to give the same from run settings
```
<RunSettings>
	<RunConfiguration>
		<Loggers>
			<Logger>trx</Logger>
			<Logger>blame;key1=value1</Logger>
		</Loggers>
	</RunConfiguration>
</RunSettings>
```
Loggers from both command line and run settings are merged.

**Change 2**:
Loggers are initialized in design mode also from now. 

**Scenarios validated**:
1. Loggers are working from IDE and command line
2. Loggers given from command line and run settings are merged.
3. Invalid loggers and logger element are handled properly
4. In case of parsing error, warning is coming.
5. Same logger given multiple times is taken only once.

**Things which we are not taking as part of this PR**:
1. There is a [bug](https://github.com/Microsoft/vstest/issues/1343) in TPv2 that console logger initializes in IDE scenarios also. 
2. <Logger>console</Logger> cant be given from runsettings. But /logger:console can be given from command line.

**Open questions**:
**1. Should we add <Logger>console</Logger>  support in runsettings?
2. Does Loggers and logger tag names good? OR is there any better suggestion.**

### CODE REVIEW

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 1]
```diff
@@ -286,5 +276,33 @@ private static void AddExtensionAssembliesFromExtensionDirectory()
...
+                var parseSucceeded = ParserUtilities.TryParseLoggerArgument(logger, out loggerIdentifier, out parameters);
+
+                if (parseSucceeded)
+                {
+                    TestLoggerManager.Instance.UpdateLoggerList(logger, loggerIdentifier, parameters);
+                }
+                else
+                {
+                    if (EqtTrace.IsWarningEnabled)
+                        EqtTrace.Warning(string.Format("The Test Logger URI &apos;{0}&apos; is not valid.  The Test Logger will be ignored.", logger));
```
Comment (abhishekkumawat23) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: I checked that localization is not done in warning case, Do I need to add it?

**Comment by codito** [comment]
This is a new feature. It requires a RFC in vstest-docs. Example of RFCs: https://github.com/Microsoft/vstest-docs/tree/master/RFCs. Please call out motivation, specification (if logger is specified in command line and runsettings, which one wins?) etc. in the doc.

Suggest keeping `Loggers` outside `RunConfiguration`. Example below. Rationale:

0. Consistency. Data collection [configuration](https://github.com/Microsoft/vstest-docs/blob/master/docs/configure.md#section-2-data-collection)
1. `RunConfiguration` is immutable and not extensible by any test platform extensions. 3rd party specification in runsettings may lead to unnecessary errors in core workflow.
2. TP passes runsettings to each extension, allowing them to provide their own settings (similar to DataCollector). This spec here should be open for extension. Please consider scenarios: (a) where a logger requires more settings than key/value, (b) custom settings (similar to code coverage datacollector)
3. What's the rationale for using `;` as a separator for runsettings? Why not provide `<logger name="blame"><key>value</key></logger>`? XML can be much more expressive, same constraints as CLI parameters wouldn't apply, right?

Example:
```
<RunSettings>
	<RunConfiguration>
	</RunConfiguration>
	<Loggers>
		<Logger>trx</Logger>
		<Logger name="blame" enabled="true">
                      <key>Value</key>
                </Logger>
	</Loggers>
</RunSettings>
```

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs (hunk 1 above)
Comment (codito) [reply src/Microsoft.TestPlatform.Client/TestPlatform.cs]: EqtTrace is purely for diagnostic purposes. No loc is required.

Doesn't `Warning` take string params? We may consider string interpolation for readability.

File: src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs]: Within namespace.

File: src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+using System.Collections.Generic;
+
+namespace Microsoft.VisualStudio.TestPlatform.Common.Utilities
+{
+    /// <summary>
+    /// Utility methods for Parsing arguments via command line or run settings.
+    /// </summary>
+    public static class ParserUtilities
```
Comment (acesiddhu) [inline src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs]: >ParserUtilities [](start = 24, length = 15)

call it ArgumentParser

File: src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs (hunk 2 above)
Comment (abhishekkumawat23) [reply src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs]: done

File: src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs (hunk 3 above)
Comment (abhishekkumawat23) [reply src/Microsoft.TestPlatform.Common/Utilities/ParserUtilities.cs]: done

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs (hunk 1 above)
Comment (abhishekkumawat23) [reply src/Microsoft.TestPlatform.Client/TestPlatform.cs]: done

**Comment by pvlakshm** [comment]
Yes. Note that we should have an RFC for all new features we look to add, please.

File: src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    /// </summary>
+    public static class ArgumentParser
+    {
+        /// <summary>
+        /// Parses the parameters passed as name values pairs along with the logger argument.
+        /// </summary>
+        /// <param name="argument">Logger argument</param>
+        /// <param name="loggerIdentifier">Receives logger Uri or friendly name.</param>
+        /// <param name="parameters">Receives parse name value pairs.</param>
+        /// <returns>True is successful, false otherwise.</returns>
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs]: can you please provide a sample input Argument?

File: src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            char[] ArgumentSeperator = new char[] { ';' };
+            char[] NameValueSeperator = new char[] { '=' };
+
+            var argumentParts = argument.Split(ArgumentSeperator, StringSplitOptions.RemoveEmptyEntries);
+
+            if (argumentParts.Length > 0 && !argumentParts[0].Contains("="))
+            {
+                loggerIdentifier = argumentParts[0];
+
+                if (argumentParts.Length > 1)
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs]: shouldn't the top if check if replaced with argumentParts.Length > 1, as if it is = 1 you will set parseSuccess to false

File: src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        /// <param name="loggerIdentifier">Receives logger Uri or friendly name.</param>
+        /// <param name="parameters">Receives parse name value pairs.</param>
+        /// <returns>True is successful, false otherwise.</returns>
+        public static bool TryParseLoggerArgument(string argument, out string loggerIdentifier, out Dictionary<string, string> parameters)
+        {
+            loggerIdentifier = null;
+            parameters = null;
+
+            var parseSucceeded = true;
+            char[] ArgumentSeperator = new char[] { ';' };
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs]: move them as const, & out of this method

File: src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        /// <param name="argument">Logger argument</param>
+        /// <param name="loggerIdentifier">Receives logger Uri or friendly name.</param>
+        /// <param name="parameters">Receives parse name value pairs.</param>
+        /// <returns>True is successful, false otherwise.</returns>
+        public static bool TryParseLoggerArgument(string argument, out string loggerIdentifier, out Dictionary<string, string> parameters)
+        {
+            loggerIdentifier = null;
+            parameters = null;
+
+            var parseSucceeded = true;
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/Utilities/ArgumentParser.cs]: kinda strange, you are staring with assumption that parseSucceeded = true, & setting it to false if it fails, what if you start it as false, then you can save last else statement

File: src/Microsoft.TestPlatform.ObjectModel/RunSettings/RunConfiguration.cs [hunk 8]
```diff
@@ -553,6 +564,18 @@ public override XmlElement ToXml()
                 root.AppendChild(testAdaptersPaths);
             }
 
+            if(this.Loggers != null && this.Loggers.Count > 0)
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.ObjectModel/RunSettings/RunConfiguration.cs]: please change .Count to .Any()
do we need a mull check, can Loggers be null?

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 9]
```diff
@@ -286,5 +276,33 @@ private static void AddExtensionAssembliesFromExtensionDirectory()
                 TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths;
             }
         }
+
+        /// <summary>
+        /// Update the loggers from run settings
+        /// </summary>
+        /// <param name="runSettings"></param>
+        private void UpdateLoggersFromRunSettings(string runSettings)
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: Tests needed for this

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 10]
```diff
@@ -141,6 +140,62 @@ public void ResetOptions()
...
+                testLoggerManager.AddLogger(new ConsoleLogger(), ConsoleLogger.ExtensionUri, null);
+            }
+
+        }
+
+        /// <summary>
+        /// Update loggers from run settings.
+        /// </summary>
+        /// <param name="runSettings"></param>
+        private void UpdateLoggersFromRunSettings(string runSettings, TestLoggerManager testLoggerManager)
```
Comment (cltshivash) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: >UpdateLoggersFromRunSettings [](start = 21, length = 28)

why can't this move into the the loggermanager

File: src/Microsoft.TestPlatform.ObjectModel/Logger/LoggerRunSettings.cs [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    public class LoggerRunSettings : TestRunSettings
+    {
+        private string loggerRootName = string.Empty;
+        private string loggersName = string.Empty;
+        private string loggerName = string.Empty;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LoggerRunSettings"/> class.
+        /// </summary>
+        public LoggerRunSettings() : base(Constants.DataCollectionRunSettingsName)
```
Comment (cltshivash) [inline src/Microsoft.TestPlatform.ObjectModel/Logger/LoggerRunSettings.cs]: fix this

File: src/vstest.console/Processors/EnableLoggerArgumentProcessor.cs [hunk 12]
```diff
@@ -144,31 +148,79 @@ public EnableLoggerArgumentExecutor(TestLoggerManager loggerManager)
...
+            var loggerRunSettings = XmlRunSettingsUtilities.GetLoggerRunSettings(settings) ?? new LoggerRunSettings();
 
-                if (parseSucceeded)
+            string loggerIdentifier = null;
+            Dictionary<string, string> parameters = null;
+            var parseSucceeded = LoggerUtilities.TryParseLoggerArgument(argument, out loggerIdentifier, out parameters);
+
+            if (parseSucceeded)
+            {
+                if (!DoesLoggerSettingExist(loggerIdentifier, loggerRunSettings, out var loggerSetting))
```
Comment (cltshivash) [inline src/vstest.console/Processors/EnableLoggerArgumentProcessor.cs]: cmdline should take precedence...

**Pull request closed without merging**